File: desktop-apps/macos/Sources/FriendManager.swift
import Foundation
import Network

// MARK: - 枚举定义
public enum FriendStatus: String, CaseIterable, Codable {
    case pending = "pending"
    case accepted = "accepted"
    case blocked = "blocked"
    case declined = "declined"
}

public enum OnlineStatus: String, CaseIterable, Codable {
    case online = "online"
    case offline = "offline"
    case away = "away"
    case busy = "busy"
}

// MARK: - 数据模型
public struct Friend: Codable, Identifiable {
    public let id: String
    public let username: String
    public let email: String?
    public let avatarUrl: String?
    public let status: FriendStatus
    public let onlineStatus: OnlineStatus
    public let lastSeen: String?
    public let createdAt: String
    public let updatedAt: String
    
    enum CodingKeys: String, CodingKey {
        case id, username, email, status, createdAt, updatedAt
        case avatarUrl = "avatar_url"
        case onlineStatus = "online_status"
        case lastSeen = "last_seen"
    }
}

public struct FriendRequest: Codable, Identifiable {
    public let id: String
    public let fromUserId: String
    public let toUserId: String
    public let fromUsername: String
    public let toUsername: String
    public let message: String?
    public let status: FriendStatus
    public let createdAt: String
    public let updatedAt: String
    
    enum CodingKeys: String, CodingKey {
        case id, message, status, createdAt, updatedAt
        case fromUserId = "from_user_id"
        case toUserId = "to_user_id"
        case fromUsername = "from_username"
        case toUsername = "to_username"
    }
}

public struct AddFriendRequest: Codable {
    public let username: String
    public let message: String?
    
    public init(username: String, message: String? = nil) {
        self.username = username
        self.message = message
    }
}

public struct RespondToRequestData: Codable {
    public let action: String
    
    public init(accept: Bool) {
        self.action = accept ? "accept" : "decline"
    }
}

public struct UpdateStatusRequest: Codable {
    public let status: String
    
    public init(status: OnlineStatus) {
        self.status = status.rawValue
    }
}

public struct FriendResponse: Codable {
    public let success: Bool
    public let message: String?
    public let friends: [Friend]?
    public let requests: [FriendRequest]?
    
    public init(success: Bool, message: String? = nil, friends: [Friend]? = nil, requests: [FriendRequest]? = nil) {
        self.success = success
        self.message = message
        self.friends = friends
        self.requests = requests
    }
}

// MARK: - 错误定义
public enum FriendManagerError: Error, LocalizedError {
    case notAuthenticated
    case networkError(Error)
    case invalidResponse
    case serverError(String)
    case friendNotFound
    case requestNotFound
    
    public var errorDescription: String? {
        switch self {
        case .notAuthenticated:
            return "用户未认证"
        case .networkError(let error):
            return "网络错误: \(error.localizedDescription)"
        case .invalidResponse:
            return "无效的服务器响应"
        case .serverError(let message):
            return "服务器错误: \(message)"
        case .friendNotFound:
            return "好友未找到"
        case .requestNotFound:
            return "好友请求未找到"
        }
    }
}

// MARK: - 事件回调协议
public protocol FriendManagerDelegate: AnyObject {
    func friendManager(_ manager: FriendManager, didUpdateFriends friends: [Friend])
    func friendManager(_ manager: FriendManager, didUpdateRequests requests: [FriendRequest])
    func friendManager(_ manager: FriendManager, didReceiveNewRequest request: FriendRequest)
    func friendManager(_ manager: FriendManager, didUpdateFriendStatus friend: Friend)
    func friendManager(_ manager: FriendManager, didEncounterError error: FriendManagerError)
}

// MARK: - 好友管理器
@MainActor
public class FriendManager: ObservableObject {
    // MARK: - 属性
    private let httpClient: HttpClient
    private let authManager: AuthManager
    
    @Published public private(set) var friends: [Friend] = []
    @Published public private(set) var pendingRequests: [FriendRequest] = []
    @Published public private(set) var currentOnlineStatus: OnlineStatus = .offline
    @Published public private(set) var isLoading = false
    
    public weak var delegate: FriendManagerDelegate?
    
    private var serverUrl: String
    private var autoRefreshEnabled = true
    private var refreshInterval: TimeInterval = 30.0
    private var refreshTimer: Timer?
    private var networkMonitor: NWPathMonitor?
    
    // MARK: - 初始化
    public init(httpClient: HttpClient, authManager: AuthManager, serverUrl: String = "http://localhost:8081") {
        self.httpClient = httpClient
        self.authManager = authManager
        self.serverUrl = serverUrl
        
        setupNetworkMonitoring()
        startAutoRefresh()
    }
    
    deinit {
        stopAutoRefresh()
        networkMonitor?.cancel()
    }
    
    // MARK: - 网络监控
    private func setupNetworkMonitoring() {
        networkMonitor = NWPathMonitor()
        networkMonitor?.pathUpdateHandler = { [weak self] path in
            DispatchQueue.main.async {
                if path.status == .satisfied {
                    self?.refreshFriendData()
                }
            }
        }
        
        let queue = DispatchQueue(label: "NetworkMonitor")
        networkMonitor?.start(queue: queue)
    }
    
    // MARK: - 自动刷新
    private func startAutoRefresh() {
        guard autoRefreshEnabled else { return }
        
        refreshTimer = Timer.scheduledTimer(withTimeInterval: refreshInterval, repeats: true) { [weak self] _ in
            Task {
                await self?.refreshFriendData()
            }
        }
    }
    
    private func stopAutoRefresh() {
        refreshTimer?.invalidate()
        refreshTimer = nil
    }
    
    public func setAutoRefresh(enabled: Bool, interval: TimeInterval = 30.0) {
        autoRefreshEnabled = enabled
        refreshInterval = interval
        
        stopAutoRefresh()
        if enabled {
            startAutoRefresh()
        }
    }
    
    // MARK: - 好友列表管理
    public func getFriendList() async throws -> [Friend] {
        guard authManager.isLoggedIn else {
            throw FriendManagerError.notAuthenticated
        }
        
        isLoading = true
        defer { isLoading = false }
        
        do {
            let url = buildUrl(endpoint: "/api/friends")
            let request = createAuthenticatedRequest(url: url, method: "GET")
            let response = try await httpClient.sendRequest(request)
            
            if response.statusCode == 200 {
                let friendResponse = try JSONDecoder().decode(FriendResponse.self, from: response.data)
                
                if friendResponse.success {
                    let newFriends = friendResponse.friends ?? []
                    self.friends = newFriends
                    delegate?.friendManager(self, didUpdateFriends: newFriends)
                    return newFriends
                } else {
                    throw FriendManagerError.serverError(friendResponse.message ?? "获取好友列表失败")
                }
            } else {
                throw FriendManagerError.serverError("HTTP \(response.statusCode)")
            }
        } catch {
            let friendError = error as? FriendManagerError ?? FriendManagerError.networkError(error)
            delegate?.friendManager(self, didEncounterError: friendError)
            throw friendError
        }
    }
    
    public func getFriendRequests() async throws -> [FriendRequest] {
        guard authManager.isLoggedIn else {
            throw FriendManagerError.notAuthenticated
        }
        
        isLoading = true
        defer { isLoading = false }
        
        do {
            let url = buildUrl(endpoint: "/api/friends/requests")
            let request = createAuthenticatedRequest(url: url, method: "GET")
            let response = try await httpClient.sendRequest(request)
            
            if response.statusCode == 200 {
                let friendResponse = try JSONDecoder().decode(FriendResponse.self, from: response.data)
                
                if friendResponse.success {
                    let newRequests = friendResponse.requests ?? []
                    self.pendingRequests = newRequests
                    delegate?.friendManager(self, didUpdateRequests: newRequests)
                    return newRequests
                } else {
                    throw FriendManagerError.serverError(friendResponse.message ?? "获取好友请求失败")
                }
            } else {
                throw FriendManagerError.serverError("HTTP \(response.statusCode)")
            }
        } catch {
            let friendError = error as? FriendManagerError ?? FriendManagerError.networkError(error)
            delegate?.friendManager(self, didEncounterError: friendError)
            throw friendError
        }
    }
    
    // MARK: - 好友操作
    public func addFriend(username: String, message: String? = nil) async throws -> Bool {
        guard authManager.isLoggedIn else {
            throw FriendManagerError.notAuthenticated
        }
        
        isLoading = true
        defer { isLoading = false }
        
        do {
            let addRequest = AddFriendRequest(username: username, message: message)
            let requestData = try JSONEncoder().encode(addRequest)
            
            let url = buildUrl(endpoint: "/api/friends/add")
            let request = createAuthenticatedRequest(url: url, method: "POST", body: requestData)
            let response = try await httpClient.sendRequest(request)
            
            if response.statusCode == 200 || response.statusCode == 201 {
                let friendResponse = try JSONDecoder().decode(FriendResponse.self, from: response.data)
                
                if friendResponse.success {
                    // 刷新好友请求列表
                    try await getFriendRequests()
                    return true
                } else {
                    throw FriendManagerError.serverError(friendResponse.message ?? "添加好友失败")
                }
            } else {
                throw FriendManagerError.serverError("HTTP \(response.statusCode)")
            }
        } catch {
            let friendError = error as? FriendManagerError ?? FriendManagerError.networkError(error)
            delegate?.friendManager(self, didEncounterError: friendError)
            throw friendError
        }
    }
    
    public func respondToFriendRequest(requestId: String, accept: Bool) async throws -> Bool {
        guard authManager.isLoggedIn else {
            throw FriendManagerError.notAuthenticated
        }
        
        isLoading = true
        defer { isLoading = false }
        
        do {
            let respondData = RespondToRequestData(accept: accept)
            let requestData = try JSONEncoder().encode(respondData)
            
            let url = buildUrl(endpoint: "/api/friends/requests/\(requestId)")
            let request = createAuthenticatedRequest(url: url, method: "PUT", body: requestData)
            let response = try await httpClient.sendRequest(request)
            
            if response.statusCode == 200 {
                let friendResponse = try JSONDecoder().decode(FriendResponse.self, from: response.data)
                
                if friendResponse.success {
                    // 刷新好友列表和请求列表
                    async let friendsTask = getFriendList()
                    async let requestsTask = getFriendRequests()
                    
                    _ = try await [friendsTask, requestsTask]
                    return true
                } else {
                    throw FriendManagerError.serverError(friendResponse.message ?? "响应好友请求失败")
                }
            } else {
                throw FriendManagerError.serverError("HTTP \(response.statusCode)")
            }
        } catch {
            let friendError = error as? FriendManagerError ?? FriendManagerError.networkError(error)
            delegate?.friendManager(self, didEncounterError: friendError)
            throw friendError
        }
    }
    
    public func removeFriend(friendId: String) async throws -> Bool {
        guard authManager.isLoggedIn else {
            throw FriendManagerError.notAuthenticated
        }
        
        isLoading = true
        defer { isLoading = false }
        
        do {
            let url = buildUrl(endpoint: "/api/friends/\(friendId)")
            let request = createAuthenticatedRequest(url: url, method: "DELETE")
            let response = try await httpClient.sendRequest(request)
            
            if response.statusCode == 200 {
                let friendResponse = try JSONDecoder().decode(FriendResponse.self, from: response.data)
                
                if friendResponse.success {
                    // 从本地列表中移除好友
                    friends.removeAll { $0.id == friendId }
                    delegate?.friendManager(self, didUpdateFriends: friends)
                    return true
                } else {
                    throw FriendManagerError.serverError(friendResponse.message ?? "删除好友失败")
                }
            } else {
                throw FriendManagerError.serverError("HTTP \(response.statusCode)")
            }
        } catch {
            let friendError = error as? FriendManagerError ?? FriendManagerError.networkError(error)
            delegate?.friendManager(self, didEncounterError: friendError)
            throw friendError
        }
    }
    
    public func blockFriend(friendId: String) async throws -> Bool {
        return try await updateFriendBlockStatus(friendId: friendId, block: true)
    }
    
    public func unblockFriend(friendId: String) async throws -> Bool {
        return try await updateFriendBlockStatus(friendId: friendId, block: false)
    }
    
    private func updateFriendBlockStatus(friendId: String, block: Bool) async throws -> Bool {
        guard authManager.isLoggedIn else {
            throw FriendManagerError.notAuthenticated
        }
        
        isLoading = true
        defer { isLoading = false }
        
        do {
            let action = block ? "block" : "unblock"
            let actionData = ["action": action]
            let requestData = try JSONSerialization.data(withJSONObject: actionData)
            
            let endpoint = block ? "/api/friends/\(friendId)/block" : "/api/friends/\(friendId)/unblock"
            let url = buildUrl(endpoint: endpoint)
            let request = createAuthenticatedRequest(url: url, method: "PUT", body: requestData)
            let response = try await httpClient.sendRequest(request)
            
            if response.statusCode == 200 {
                let friendResponse = try JSONDecoder().decode(FriendResponse.self, from: response.data)
                
                if friendResponse.success {
                    // 刷新好友列表
                    try await getFriendList()
                    return true
                } else {
                    let actionText = block ? "屏蔽" : "取消屏蔽"
                    throw FriendManagerError.serverError(friendResponse.message ?? "\(actionText)好友失败")
                }
            } else {
                throw FriendManagerError.serverError("HTTP \(response.statusCode)")
            }
        } catch {
            let friendError = error as? FriendManagerError ?? FriendManagerError.networkError(error)
            delegate?.friendManager(self, didEncounterError: friendError)
            throw friendError
        }
    }
    
    // MARK: - 在线状态管理
    public func updateOnlineStatus(_ status: OnlineStatus) async throws {
        guard authManager.isLoggedIn else {
            throw FriendManagerError.notAuthenticated
        }
        
        do {
            let statusRequest = UpdateStatusRequest(status: status)
            let requestData = try JSONEncoder().encode(statusRequest)
            
            let url = buildUrl(endpoint: "/api/user/status")
            let request = createAuthenticatedRequest(url: url, method: "PUT", body: requestData)
            let response = try await httpClient.sendRequest(request)
            
            if response.statusCode == 200 {
                currentOnlineStatus = status
            } else {
                throw FriendManagerError.serverError("HTTP \(response.statusCode)")
            }
        } catch {
            let friendError = error as? FriendManagerError ?? FriendManagerError.networkError(error)
            delegate?.friendManager(self, didEncounterError: friendError)
            throw friendError
        }
    }
    
    // MARK: - 查找方法
    public func findFriend(byId friendId: String) -> Friend? {
        return friends.first { $0.id == friendId }
    }
    
    public func findFriend(byUsername username: String) -> Friend? {
        return friends.first { $0.username == username }
    }
    
    public func findRequest(byId requestId: String) -> FriendRequest? {
        return pendingRequests.first { $0.id == requestId }
    }
    
    // MARK: - 配置方法
    public func setServerUrl(_ url: String) {
        serverUrl = url
    }
    
    // MARK: - 数据刷新
    public func refreshFriendData() async {
        guard authManager.isLoggedIn else { return }
        
        do {
            async let friendsTask = getFriendList()
            async let requestsTask = getFriendRequests()
            
            _ = try await [friendsTask, requestsTask]
        } catch {
            // 错误已经在各自的方法中处理
        }
    }
    
    // MARK: - 辅助方法
    private func buildUrl(endpoint: String) -> String {
        return serverUrl + endpoint
    }
    
    private func createAuthenticatedRequest(url: String, method: String, body: Data? = nil) -> HttpRequest {
        var request = HttpRequest(
            url: url,
            method: method,
            headers: ["Content-Type": "application/json"],
            body: body
        )
        
        // 添加认证头
        if let token = authManager.accessToken {
            request.headers["Authorization"] = "Bearer \(token)"
        }
        
        return request
    }
}

// MARK: - 全局实例管理
public class FriendManagerFactory {
    public static let shared = FriendManagerFactory()
    private var _instance: FriendManager?
    
    private init() {}
    
    public func initialize(httpClient: HttpClient, authManager: AuthManager, serverUrl: String = "http://localhost:8081") {
        _instance = FriendManager(httpClient: httpClient, authManager: authManager, serverUrl: serverUrl)
    }
    
    public var instance: FriendManager? {
        return _instance
    }
    
    public func cleanup() {
        _instance = nil
    }
}

File: desktop-apps/windows/src/friend_manager.cpp
#include "friend_manager.h"
#include <json/json.h>
#include <thread>
#include <chrono>
#include <sstream>

// 全局好友管理器实例
std::unique_ptr<FriendManager> g_friendManager = nullptr;

// 全局初始化和清理函数
bool initFriendManager() {
    if (g_friendManager) {
        return true; // 已经初始化
    }
    
    g_friendManager = std::make_unique<FriendManager>();
    return g_friendManager->initialize();
}

void cleanupFriendManager() {
    if (g_friendManager) {
        g_friendManager->cleanup();
        g_friendManager.reset();
    }
}

// Friend 结构实现
Friend::Friend(const std::string& json) {
    Json::Value root;
    Json::Reader reader;
    
    if (reader.parse(json, root)) {
        id = root.get("id", "").asString();
        username = root.get("username", "").asString();
        email = root.get("email", "").asString();
        avatarUrl = root.get("avatar_url", "").asString();
        status = stringToFriendStatus(root.get("status", "").asString());
        onlineStatus = stringToOnlineStatus(root.get("online_status", "").asString());
        lastSeen = root.get("last_seen", "").asString();
        createdAt = root.get("created_at", "").asString();
        updatedAt = root.get("updated_at", "").asString();
    }
}

std::string Friend::toJson() const {
    Json::Value root;
    root["id"] = id;
    root["username"] = username;
    root["email"] = email;
    root["avatar_url"] = avatarUrl;
    root["status"] = friendStatusToString(status);
    root["online_status"] = onlineStatusToString(onlineStatus);
    root["last_seen"] = lastSeen;
    root["created_at"] = createdAt;
    root["updated_at"] = updatedAt;
    
    Json::StreamWriterBuilder builder;
    return Json::writeString(builder, root);
}

// FriendRequest 结构实现
FriendRequest::FriendRequest(const std::string& json) {
    Json::Value root;
    Json::Reader reader;
    
    if (reader.parse(json, root)) {
        id = root.get("id", "").asString();
        fromUserId = root.get("from_user_id", "").asString();
        toUserId = root.get("to_user_id", "").asString();
        fromUsername = root.get("from_username", "").asString();
        toUsername = root.get("to_username", "").asString();
        message = root.get("message", "").asString();
        status = stringToFriendStatus(root.get("status", "").asString());
        createdAt = root.get("created_at", "").asString();
        updatedAt = root.get("updated_at", "").asString();
    }
}

std::string FriendRequest::toJson() const {
    Json::Value root;
    root["id"] = id;
    root["from_user_id"] = fromUserId;
    root["to_user_id"] = toUserId;
    root["from_username"] = fromUsername;
    root["to_username"] = toUsername;
    root["message"] = message;
    root["status"] = friendStatusToString(status);
    root["created_at"] = createdAt;
    root["updated_at"] = updatedAt;
    
    Json::StreamWriterBuilder builder;
    return Json::writeString(builder, root);
}

// AddFriendRequest 结构实现
std::string AddFriendRequest::toJson() const {
    Json::Value root;
    root["username"] = username;
    root["message"] = message;
    
    Json::StreamWriterBuilder builder;
    return Json::writeString(builder, root);
}

// FriendResponse 结构实现
FriendResponse::FriendResponse(const std::string& json) {
    Json::Value root;
    Json::Reader reader;
    
    if (reader.parse(json, root)) {
        success = root.get("success", false).asBool();
        message = root.get("message", "").asString();
        
        const Json::Value& friendsArray = root["friends"];
        if (friendsArray.isArray()) {
            for (const auto& friendJson : friendsArray) {
                Json::StreamWriterBuilder builder;
                std::string friendStr = Json::writeString(builder, friendJson);
                friends.emplace_back(friendStr);
            }
        }
        
        const Json::Value& requestsArray = root["requests"];
        if (requestsArray.isArray()) {
            for (const auto& requestJson : requestsArray) {
                Json::StreamWriterBuilder builder;
                std::string requestStr = Json::writeString(builder, requestJson);
                requests.emplace_back(requestStr);
            }
        }
    }
}

// FriendManager 实现
FriendManager::FriendManager() 
    : serverUrl("http://localhost:8081")
    , currentOnlineStatus(OnlineStatus::OFFLINE)
    , autoRefreshEnabled(true)
    , refreshInterval(30)
    , isRunning(false) {
}

FriendManager::~FriendManager() {
    cleanup();
}

bool FriendManager::initialize() {
    httpClient = std::make_unique<HttpClient>();
    if (!httpClient) {
        return false;
    }
    
    currentOnlineStatus = OnlineStatus::ONLINE;
    isRunning = true;
    
    if (autoRefreshEnabled) {
        startAutoRefresh();
    }
    
    return true;
}

void FriendManager::cleanup() {
    isRunning = false;
    stopAutoRefresh();
    
    if (refreshThread.joinable()) {
        refreshThread.join();
    }
    
    httpClient.reset();
}

// 同步方法实现
FriendResponse FriendManager::getFriendList() {
    std::string url = buildUrl("/api/friends");
    HttpRequest request = createAuthenticatedRequest(url, "GET");
    
    HttpResponse response = httpClient->get(url);
    return parseResponse(response);
}

FriendResponse FriendManager::getFriendRequests() {
    std::string url = buildUrl("/api/friends/requests");
    HttpRequest request = createAuthenticatedRequest(url, "GET");
    
    HttpResponse response = httpClient->get(url);
    return parseResponse(response);
}

bool FriendManager::addFriend(const std::string& username, const std::string& message) {
    AddFriendRequest request;
    request.username = username;
    request.message = message;
    
    std::string url = buildUrl("/api/friends/add");
    std::string jsonData = request.toJson();
    
    HttpResponse response = httpClient->post(url, jsonData);
    
    if (response.statusCode == 200 || response.statusCode == 201) {
        FriendResponse friendResponse(response.body);
        return friendResponse.success;
    }
    
    return false;
}

bool FriendManager::respondToFriendRequest(const std::string& requestId, bool accept) {
    Json::Value root;
    root["action"] = accept ? "accept" : "decline";
    
    Json::StreamWriterBuilder builder;
    std::string jsonData = Json::writeString(builder, root);
    
    std::string url = buildUrl("/api/friends/requests/" + requestId);
    HttpResponse response = httpClient->put(url, jsonData);
    
    if (response.statusCode == 200) {
        FriendResponse friendResponse(response.body);
        return friendResponse.success;
    }
    
    return false;
}

bool FriendManager::removeFriend(const std::string& friendId) {
    std::string url = buildUrl("/api/friends/" + friendId);
    HttpResponse response = httpClient->deleteRequest(url);
    
    if (response.statusCode == 200) {
        FriendResponse friendResponse(response.body);
        return friendResponse.success;
    }
    
    return false;
}

bool FriendManager::blockFriend(const std::string& friendId) {
    Json::Value root;
    root["action"] = "block";
    
    Json::StreamWriterBuilder builder;
    std::string jsonData = Json::writeString(builder, root);
    
    std::string url = buildUrl("/api/friends/" + friendId + "/block");
    HttpResponse response = httpClient->put(url, jsonData);
    
    if (response.statusCode == 200) {
        FriendResponse friendResponse(response.body);
        return friendResponse.success;
    }
    
    return false;
}

bool FriendManager::unblockFriend(const std::string& friendId) {
    Json::Value root;
    root["action"] = "unblock";
    
    Json::StreamWriterBuilder builder;
    std::string jsonData = Json::writeString(builder, root);
    
    std::string url = buildUrl("/api/friends/" + friendId + "/unblock");
    HttpResponse response = httpClient->put(url, jsonData);
    
    if (response.statusCode == 200) {
        FriendResponse friendResponse(response.body);
        return friendResponse.success;
    }
    
    return false;
}

// 异步方法实现
void FriendManager::getFriendListAsync(FriendListCallback callback) {
    std::thread([this, callback]() {
        FriendResponse response = getFriendList();
        if (callback) {
            callback(response);
        }
    }).detach();
}

void FriendManager::getFriendRequestsAsync(FriendRequestCallback callback) {
    std::thread([this, callback]() {
        FriendResponse response = getFriendRequests();
        if (callback) {
            callback(response);
        }
    }).detach();
}

void FriendManager::addFriendAsync(const std::string& username, const std::string& message, AddFriendCallback callback) {
    std::thread([this, username, message, callback]() {
        bool success = addFriend(username, message);
        if (callback) {
            callback(success, success ? "Friend request sent successfully" : "Failed to send friend request");
        }
    }).detach();
}

void FriendManager::respondToFriendRequestAsync(const std::string& requestId, bool accept, RespondFriendCallback callback) {
    std::thread([this, requestId, accept, callback]() {
        bool success = respondToFriendRequest(requestId, accept);
        std::string message = success ? 
            (accept ? "Friend request accepted" : "Friend request declined") : 
            "Failed to respond to friend request";
        if (callback) {
            callback(success, message);
        }
    }).detach();
}

void FriendManager::removeFriendAsync(const std::string& friendId, RemoveFriendCallback callback) {
    std::thread([this, friendId, callback]() {
        bool success = removeFriend(friendId);
        if (callback) {
            callback(success, success ? "Friend removed successfully" : "Failed to remove friend");
        }
    }).detach();
}

void FriendManager::blockFriendAsync(const std::string& friendId, RespondFriendCallback callback) {
    std::thread([this, friendId, callback]() {
        bool success = blockFriend(friendId);
        if (callback) {
            callback(success, success ? "Friend blocked successfully" : "Failed to block friend");
        }
    }).detach();
}

void FriendManager::unblockFriendAsync(const std::string& friendId, RespondFriendCallback callback) {
    std::thread([this, friendId, callback]() {
        bool success = unblockFriend(friendId);
        if (callback) {
            callback(success, success ? "Friend unblocked successfully" : "Failed to unblock friend");
        }
    }).detach();
}

// 好友状态管理
void FriendManager::updateOnlineStatus(OnlineStatus status) {
    currentOnlineStatus = status;
    
    Json::Value root;
    root["status"] = onlineStatusToString(status);
    
    Json::StreamWriterBuilder builder;
    std::string jsonData = Json::writeString(builder, root);
    
    std::string url = buildUrl("/api/user/status");
    httpClient->put(url, jsonData);
}

OnlineStatus FriendManager::getOnlineStatus() const {
    return currentOnlineStatus;
}

Friend* FriendManager::findFriendById(const std::string& friendId) {
    for (auto& friend_ : friends) {
        if (friend_.id == friendId) {
            return &friend_;
        }
    }
    return nullptr;
}

Friend* FriendManager::findFriendByUsername(const std::string& username) {
    for (auto& friend_ : friends) {
        if (friend_.username == username) {
            return &friend_;
        }
    }
    return nullptr;
}

// 配置方法
void FriendManager::setServerUrl(const std::string& url) {
    serverUrl = url;
}

void FriendManager::enableAutoRefresh(bool enable) {
    autoRefreshEnabled = enable;
    if (enable && isRunning) {
        startAutoRefresh();
    } else {
        stopAutoRefresh();
    }
}

void FriendManager::setRefreshInterval(int seconds) {
    refreshInterval = seconds;
}

// 事件回调设置
void FriendManager::setOnFriendListUpdated(FriendListCallback callback) {
    onFriendListUpdated = callback;
}

void FriendManager::setOnFriendRequestReceived(FriendRequestCallback callback) {
    onFriendRequestReceived = callback;
}

void FriendManager::setOnFriendStatusChanged(std::function<void(const Friend&)> callback) {
    onFriendStatusChanged = callback;
}

// 内部方法
std::string FriendManager::buildUrl(const std::string& endpoint) const {
    return serverUrl + endpoint;
}

HttpRequest FriendManager::createAuthenticatedRequest(const std::string& url, const std::string& method) const {
    HttpRequest request;
    request.url = url;
    request.method = method;
    
    // 添加认证头
    if (g_authManager && g_authManager->isLoggedIn()) {
        std::string token = g_authManager->getAccessToken();
        if (!token.empty()) {
            request.headers["Authorization"] = "Bearer " + token;
        }
    }
    
    request.headers["Content-Type"] = "application/json";
    return request;
}

FriendResponse FriendManager::parseResponse(const HttpResponse& response) const {
    if (response.statusCode == 200) {
        return FriendResponse(response.body);
    }
    
    FriendResponse errorResponse;
    errorResponse.success = false;
    errorResponse.message = "HTTP Error: " + std::to_string(response.statusCode);
    return errorResponse;
}

void FriendManager::startAutoRefresh() {
    if (refreshThread.joinable()) {
        return; // 已经在运行
    }
    
    refreshThread = std::thread([this]() {
        while (isRunning && autoRefreshEnabled) {
            refreshFriendData();
            std::this_thread::sleep_for(std::chrono::seconds(refreshInterval));
        }
    });
}

void FriendManager::stopAutoRefresh() {
    // 线程会在下次循环时自动停止
}

void FriendManager::refreshFriendData() {
    // 刷新好友列表
    FriendResponse friendListResponse = getFriendList();
    if (friendListResponse.success) {
        friends = friendListResponse.friends;
        if (onFriendListUpdated) {
            onFriendListUpdated(friendListResponse);
        }
    }
    
    // 刷新好友请求
    FriendResponse requestsResponse = getFriendRequests();
    if (requestsResponse.success) {
        pendingRequests = requestsResponse.requests;
        if (onFriendRequestReceived) {
            onFriendRequestReceived(requestsResponse);
        }
    }
}

// 辅助函数实现
std::string friendStatusToString(FriendStatus status) {
    switch (status) {
        case FriendStatus::PENDING: return "pending";
        case FriendStatus::ACCEPTED: return "accepted";
        case FriendStatus::BLOCKED: return "blocked";
        case FriendStatus::DECLINED: return "declined";
        default: return "unknown";
    }
}

FriendStatus stringToFriendStatus(const std::string& status) {
    if (status == "pending") return FriendStatus::PENDING;
    if (status == "accepted") return FriendStatus::ACCEPTED;
    if (status == "blocked") return FriendStatus::BLOCKED;
    if (status == "declined") return FriendStatus::DECLINED;
    return FriendStatus::PENDING;
}

std::string onlineStatusToString(OnlineStatus status) {
    switch (status) {
        case OnlineStatus::ONLINE: return "online";
        case OnlineStatus::OFFLINE: return "offline";
        case OnlineStatus::AWAY: return "away";
        case OnlineStatus::BUSY: return "busy";
        default: return "offline";
    }
}

OnlineStatus stringToOnlineStatus(const std::string& status) {
    if (status == "online") return OnlineStatus::ONLINE;
    if (status == "offline") return OnlineStatus::OFFLINE;
    if (status == "away") return OnlineStatus::AWAY;
    if (status == "busy") return OnlineStatus::BUSY;
    return OnlineStatus::OFFLINE;
}